pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java

```diff
@@ -0,0 +1,779 @@
+                    // TODO: enable CompactionStrategy
+                    .create();
+            // TODO: schedule listen instead of foreachAndListen
+            tableview.forEachAndListen((key, value) -> handle(key, value));
+            log.debug("Successfully started the channel tableview.");
+
+            // TODO: schedule cleanupTasks by brokerRegistry
+            pulsar.getLocalMetadataStore().registerSessionListener(this::handleMetadataSessionEvent);
+            log.debug("Successfully registered the handleMetadataSessionEvent");
+
+            log.info("Successfully started the channel.");
+        } catch (Exception e) {
+            String msg = "Failed to start the channel.";
+            log.error(msg, e);
+            throw new PulsarServerException(msg, e);
+        }
+    }
+
+
+    public void close() throws PulsarServerException {
+        try {
+            if (leaderElectionService != null) {
+                leaderElectionService.close();
+                leaderElectionService = null;
+                log.debug("Successfully closed the channel leader election service.");
+            }
+
+            if (tableview != null) {
+                tableview.close();
+                tableview = null;
+                log.debug("Successfully closed the channel tableview.");
+            }
+
+            if (producer != null) {
+                producer.close();
+                producer = null;
+                log.info("Successfully closed the channel producer.");
+            }
+
+            // TODO: clean brokerRegistry
+
+            if (cleanupTasks != null) {
+                cleanupTasks.cancel(true);
+                cleanupTasks = null;
+                log.info("Successfully cancelled the cleanup tasks");
+            }
+
+            log.info("Successfully closed the channel.");
+
+        } catch (Exception e) {
+            String msg = "Failed to close the channel.";
+            log.error(msg, e);
+            throw new PulsarServerException(msg, e);
+        }
+    }
+
+    public void scheduleCompaction() throws PulsarServerException {
+        try {
+            Long threshold = pulsar.getAdminClient().topicPolicies()
+                    .getCompactionThreshold(TOPIC);
+            if (threshold == null || threshold == 0) {
+                pulsar.getAdminClient().topicPolicies()
+                        .setCompactionThreshold(TOPIC, COMPACTION_THRESHOLD);
+                log.info("Scheduled compaction on topic:{}, threshold:{} bytes", TOPIC, COMPACTION_THRESHOLD);
+            } else {
+                log.info("Already set compaction on topic:{}, threshold:{} bytes", TOPIC, COMPACTION_THRESHOLD);
+            }
+        } catch (PulsarAdminException e) {
+            throw new PulsarServerException("Failed to schedule compaction.", e);
+        }
+    }
+
+    public String getChannelOwner() {
+        CompletableFuture<Optional<LeaderBroker>> future = leaderElectionService.readCurrentLeader();
```

Comment:
I think that we are not handling concurrent access on "leaderElectionService"
in "start" we re-assign the value to the variable.
I would make it final or properly handle concurrent access

Reply:
- Added a private variable `volatile boolean isActive` and set this flag in start() and close().
- Made start() and close() synchronized.
- Added validateChannel() to validate concurrent access to `leaderElectionService, producer, and tableview` while `isActive=false`.
- Added a UT to cover this access logic.

Reply:
In fact, we need further the channel state breakdown (`boolean isActive` is not enough because `getChannelOwner()` should be accessible while creating the `producer` and `tableview` on the system topic -- the topic owner lookup will call `getChannelOwner()`.

I added the `enum StateChannel` to differentiate the channel states.

Reply:
I wonder why spotbugs is not complaining about "unconsistent sync" on `leaderElectionService`.
because we still have an access without "synchonized".
I believe that you can add a getter
`private synchronised LeaderElectionService leaderElectionService() { ... }`
and call it here

Reply:
why can't we make it "final" and assign it in the constructor, without calling "start" in the constructor?

Reply:
Made it final.

Updated.

Reply:
About the concurrent access to the member variables, still 'tableview' and 'producer' are not final.

Could you explain more why we need the synchronized access to the 'leaderElectionService' instance?
As long as it's non-null, shouldn't it be ok whether its from the old or new instance of 'leaderElectionService'?

Reply:
You are correct we have to handle correctly synchronization around tableview and producer.

The reason is related to the Java Memory Model.
If you access a field from multiple threads you are not guaranteed that each threads see the same value if you use any synchronization mechanism (there are many)

Spot bugs usually finds this problems

Reply:
Aren't the current state channel validation and synchronized start/close func enough? The invalid channel state can throw exception when the getter funcs are called during start() and close().

If this is not sufficient, can you explain why? And do you suggest forcefully synchronized access like 'synchronized getProducer()/setProducer() and synchronized getTableView()/setTableView()?

We could use ReadWriteLock to optimize this, but isn't it an overkill?